codelists/models.py

```diff
@@ -566,20 +566,50 @@ def table_with_fixed_headers(self, include_mapped_vmps=True):
 
         table_rows = self.table[1:]
         if include_mapped_vmps and self.coding_system_id == "dmd":
-            term_header_index = term_header_index or len(self.table[0])
             # ignore include_mapped_vmps if coding system is anything other than dmd
+            term_header_index = term_header_index or len(self.table[0])
             codes = [row[code_header_index] for row in table_rows]
+
             # add in mapped VMP codes
-            previous_vmps_for_this_codelist = vmp_ids_to_previous(codes)
-            for current_vmp, previous_vmp in previous_vmps_for_this_codelist:
-                assert current_vmp in codes
-                if previous_vmp not in codes:
-                    new_row = ["" for i in range(len(table_rows[0]) + 1)]
-                    new_row[code_header_index] = previous_vmp
-                    new_row[
-                        term_header_index
-                    ] = f"Mapped previous VMP for {current_vmp}"
-                    table_rows.append(new_row)
+            vmp_to_prev_mapping, mapped_vmps_for_this_codelist = vmpprev_full_mappings(
+                codes
+            )
+            # vmp_to_prev_mapping is a simple 1:1 mapping of vmp ID to the immediate previous
+            # VMP that it replaced
+            # Since we're mapping later codes as well as previous ones, we also need the
+            # reverse mapping of previous VMP to the one that replaced it
+            prev_to_vmp_mapping = {v: k for k, v in vmp_to_prev_mapping.items()}
```

Comment:
The mapping is invertible, and that assertion will pass right now.  However, I'm not convinced that it always will - it could be possible that a VMP code is replaced by 2 new codes. I'm not sure how the retrospective updating would work in that case, so maybe it never happens, but I think it's worth checking and dealing with it if it does.